### PR TITLE
Add FindLsb / FindMsb

### PR DESCRIPTION
**Pull request by fintelia**
Currently untested. I welcome feedback on whether this approach makes sense and how to go about adding tests.

### CODE REVIEW

File: src/back/msl/writer.rs [hunk 1]
```diff
@@ -1184,6 +1184,8 @@ impl<W: Write> Writer<W> {
                     Mf::ReverseBits => "reverse_bits",
                     Mf::ExtractBits => "extract_bits",
                     Mf::InsertBits => "insert_bits",
+                    Mf::FindLsb => return Err(Error::UnsupportedCall(format!("{:?}", fun))),
```
[the remark this answers is not in the document]
Comment (kvark) [reply src/back/msl/writer.rs:1187]: @fintelia ping

[the remark this answers is not in the document]
Comment (fintelia) [reply src/back/msl/writer.rs:1187]: Could you say a bit about how to inject an expression? All the other cases seem to just reference a single function

[the remark this answers is not in the document]
Comment (kvark) [reply src/back/msl/writer.rs:1187]: The `Distance` is also injecting an expression:
```rust
                    write!(self.out, "{}::abs(", NAMESPACE)?;
                    self.put_expression(arg, context, false)?;
                    write!(self.out, " - ")?;
                    self.put_expression(arg1.unwrap(), context, false)?;
                    write!(self.out, ")")?;
```

**Comment by fintelia** [comment]
Ok, it seems that passing validation requires that all GLSL inputs can be written as valid WGSL. Unfortunately, WGSL doesn't yet have `ctz` / `clz` functions so I'm not sure the right way to achieve that without simply not testing the new functionality

File: src/back/wgsl/writer.rs [hunk 2]
```diff
@@ -1563,6 +1567,16 @@ impl<W: Write> Writer<W> {
...
                     }
+                    Function::FindLsb => {
+                        write!(self.out, "((ctz(")?;
+                        self.write_expr(module, arg, func_ctx)?;
+                        write!(self.out, ") + 1) % 33) - 1")?;
+                    }
+                    Function::FindMsb => {
+                        write!(self.out, "((clz(")?;
+                        self.write_expr(module, arg, func_ctx)?;
+                        write!(self.out, ") + 1) % 33) - 1")?;
```
Comment (kvark) [inline src/back/wgsl/writer.rs:1578]: something isn't right here. The implementations for `FindLsb` and `FindMsb` look the same?

Comment (fintelia) [reply src/back/wgsl/writer.rs:1578]: Note the "t" vs. "l" in `ctz` vs. `clz`

Comment (fintelia) [reply src/back/wgsl/writer.rs:1578]: Relatedly, since these functions aren't officially added to WGSL yet someone should seriously consider other names for them that are easier to tell apart!

Comment (kvark) [reply src/back/wgsl/writer.rs:1578]: Please comment about this on the upstream issue as well!

**Comment by kvark** [comment]
I agree about the naming, it would be nice to settle this in the group.

**Comment by fintelia** [comment]
Ok, at this point it seems that the main blocker is that WGSL's builtins for this functionality haven't been specified yet. What would be the right way to push that forward so this PR can be merged?

**Comment by kvark** [comment]
If you want to use it now, we can land this with the expectation that we'll rename these methods once the group settles on them.

**Comment by kvark** [comment]
@fintelia there seem to be a small problem left on Metal?

**Comment by fintelia** [comment]
Ok, I think I managed to fix it. The issue was that there needed to be an unsigned integer -> signed integer cast, but to do that the vector width of the input needed to be known. I tried a bunch of ways to cause the type coercion to produce that without the generated code differing between them, but none ended up working. The current version checks the input width so it can insert a cast using one of `int`/`int2`/`int3`/`int4`.

File: src/back/msl/writer.rs [hunk 3]
```diff
@@ -1204,6 +1221,22 @@ impl<W: Write> Writer<W> {
                     write!(self.out, " - ")?;
                     self.put_expression(arg1.unwrap(), context, false)?;
                     write!(self.out, ")")?;
+                } else if fun == Mf::FindLsb {
+                    write!(
+                        self.out,
+                        "(((1 + int{}({}::ctz(",
+                        argument_size_suffix, NAMESPACE
+                    )?;
+                    self.put_expression(arg, context, false)?;
```
Comment (kvark) [inline src/back/msl/writer.rs:1230]: last argument should be `true` since we are inside brackets

File: src/back/msl/writer.rs [hunk 4]
```diff
@@ -1204,6 +1221,22 @@ impl<W: Write> Writer<W> {
...
+                    write!(self.out, "))) % 33) - 1)")?;
+                } else if fun == Mf::FindMsb {
+                    write!(
+                        self.out,
+                        "(((1 + int{}({}::clz(",
+                        argument_size_suffix, NAMESPACE
+                    )?;
+                    self.put_expression(arg, context, false)?;
```
Comment (kvark) [inline src/back/msl/writer.rs:1238]: also `true`

File: src/proc/typifier.rs [hunk 5]
```diff
@@ -785,6 +785,16 @@ impl<'a> ResolveContext<'a> {
                     Mf::ReverseBits |
                     Mf::ExtractBits |
                     Mf::InsertBits => res_arg.clone(),
+                    Mf::FindLsb |
+                    Mf::FindMsb => match res_arg.inner_with(types)  {
```
Comment (kvark) [inline src/proc/typifier.rs:789]: let's do `match *res_arg.inner_with(types)` like we do in other places

File: tests/in/bits.wgsl [hunk 6]
```diff
@@ -36,4 +36,20 @@ fn main() {
     u2 = extractBits(u2, 5u, 10u);
     u3 = extractBits(u3, 5u, 10u);
     u4 = extractBits(u4, 5u, 10u);
+    i = findLsb(i);
+    i2 = findLsb(i2);
+    i3 = findLsb(i3);
```
Comment (kvark) [inline tests/in/bits.wgsl:41]: why do we need to call it so many times! it's not like we are checking the result anyway, we are only interested in code generation here. So calling once should be ok.

**Review by kvark — CHANGES_REQUESTED** [review]
Thank you!

File: tests/in/bits.wgsl (hunk 6 above)
Comment (fintelia) [reply tests/in/bits.wgsl:41]: This is one call with each argument type. I can remove some if you'd like, but each one potentially generates different code.

Comment (kvark) [reply tests/in/bits.wgsl:41]: I think having something like an signed int scalar and an unsigned int vector would do (i.e. 2 cases per function).

**Review by kvark — APPROVED** [review]

**Comment by kvark** [comment]
@fintelia would you be interested to update the naming according to https://github.com/gpuweb/gpuweb/pull/2467 ?

**Comment by fintelia** [comment]
I probably wouldn't be able to get to it for a while. If anyone else wants to make the update they should go ahead!

With the functions now fully specified in WGSL, someone should probably also make sure there's conformance tests for the various edge cases. I'm not fully confident that I understood the specs fully for each of the backends; testing on lots of platforms is the only way to be sure.
